[13805] Record Replay use case documentation improvement (#328)
eProsima/Fast-DDS-docs: docs/fastdds/use_cases/rosbag_capture/rosbag_capture.rst

Add the following lines to the ``package.xml`` file after the ``buildtool_depend`` tags:

.. code-block:: bash

  <buildtool_depend>rosidl_default_generators</buildtool_depend>
  <exec_depend>rosidl_default_runtime</exec_depend>
  <member_of_group>rosidl_interface_packages</member_of_group>

The last step would be to build the package.
Run the following command within the ``fastdds_record_typesupport`` folder:

.. code-block:: bash

   RMW_IMPLEMENTATION=rmw_fastrtps_cpp colcon build

The build process will create inside the install folder a new ROS 2 overlay with all the required
libraries and scripts for the ROS 2 applications to use te type defined in the IDL file.

Fast DDS Application tuning
^^^^^^^^^^^^^^^^^^^^^^^^^^^

ROS2 adds special tokens to the topic names depending on the ROS subsystem the topic belongs to. More information
on this topic can be found `on ROS 2 design documentation
<https://design.ros2.org/articles/topic_and_service_names.html#examples-of-ros-names-to-dds-concepts>`_ .

Using the same IDL files used earlier Fast DDS-Gen can generate the required code to handle the new type in

> Fast DDS. The required changes to make so rosbag can see your application are going to be illustrated via the

Reviewer: ```suggestion
Fast DDS.
The Fast DDS application changes required for rosbag2 to communicate with it are going to be illustrated via the
```

Reviewer: This suggestion has not been applied. Avoid using `you` in the documentation.

Another suggestion could be:

```suggestion
The changes required in the Fast DDS application so rosbag2 can communicate with it are going to be illustrated via the
```

Publisher/Subscriber example generated automatically from an IDL using Fast DDS Gen. An in-depth guide to Fast DDS-Gen
can be found `here <https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/fastddsgen/fastddsgen.html>`_.

> In the case of plain topics, the namespace "rt/" is added to all topic names so it needs to be added.
> DataType names for this generated types are structured like (using the previous IDL as example)
> "fastdds_record_typesupport::idl::HelloWorld".

Reviewer: ```suggestion
In case of plain topics, the namespace "rt/" is added by ROS 2 to the DDS topic name.
DataType names for ROS 2 generated types are structured concatenating the modules names.
For the IDL being used in this example the data type name would be "fastdds_record_typesupport::idl::HelloWorld".
```

Create a new workspace different from the ROS 2 one used previously.
Copy inside the same IDL file and run Fast DDS-Gen to generate
the TypeSupport and the example source files:

.. code-block:: bash

    mkdir HelloWorldExample
    cd HelloWorldExample

>     cp ../fastdds_record_typesupport/idl/HelloWorld.idl .

Reviewer: ```suggestion
    cp <PATH_TO_ROS2_WORKSPACE>/fastdds_record_typesupport/idl/HelloWorld.idl .
```

    fastddsgen -example CMake -typeros2 HelloWorld.idl

This command will populate the current folder with the required header and source files to build the TypeSupport,
and the Publisher and Subscriber applications.

.. code-block:: shell-session

    └── HelloWorldExample
        ├── CMakeLists.txt
        ├── HelloWorld.cxx
        ├── HelloWorld.h
        ├── HelloWorld.idl
        ├── HelloWorldPublisher.cxx
        ├── HelloWorldPublisher.h
        ├── HelloWorldPubSubMain.cxx
        ├── HelloWorldPubSubTypes.cxx
        ├── HelloWorldPubSubTypes.h
        ├── HelloWorldSubscriber.cxx
        └── HelloWorldSubscriber.h

> 

Reviewer: Please remove extra blank line

The Fast DDS-Gen example should be modified taking into account the topic and type name mangling
applied by ROS 2 so communication can be established with rosbag2.

> Having used the -typeros2 Fast DDS-Gen option when generating the TypeSupport, the generated type

Reviewer: ```suggestion
Having used the ``-typeros2`` Fast DDS-Gen option when generating the TypeSupport, the generated type
```

name would already include the ROS 2 naming rule mangling.
However, the topic name must be modified manually both in the Publisher and Subscriber applications.

> Look for the create_topic command in both the HelloWorldPublisher.cxx and the HelloWorldSubscriber.cxx

Reviewer: ```suggestion
Look for the ``create_topic`` command in both the ``HelloWorldPublisher.cxx`` and the ``HelloWorldSubscriber.cxx``
```

files and modify the topic name:

.. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
    :start-after: //CREATE THE TOPIC FOR ROSBAG
    :end-before: //!
    :dedent: 4

To build this example run the following commands:

.. code-block:: bash

    mkdir build && cd build
    cmake ..
    make

This will create a HelloWorld binary file inside the build directory that can be used to launch both the Publisher
and the Subscriber applications.
Run each application in a terminal and confirm that the communication is established.

.. code-block:: bash

    ./HelloWorld publisher|subscriber

eProsima DDS Record and Play
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In order to use the generated ROS 2 TypeSupport package, the ROS 2 workspace should be sourced besides the
ROS 2 installation.
This allows rosbag2 to record the data types used in this example.
To start recording the traffic being exchanged between the Publisher/Subscriber applications the corresponding

> ROS 2 Topic
> name has to be passed to rosbag2 (not to be mistaken with the DDS Topic name).

Reviewer: ```suggestion
ROS 2 Topic name has to be passed to rosbag2 (not to be mistaken with the DDS Topic name).
```
